cpp/src/arrow/types/string.cc

```diff
@@ -24,25 +24,54 @@
 
 namespace arrow {
 
+const std::shared_ptr<DataType> BINARY(new BinaryType());
 const std::shared_ptr<DataType> STRING(new StringType());
 
+BinaryArray::BinaryArray(int32_t length, const std::shared_ptr<Buffer>& offsets,
+    const ArrayPtr& values, int32_t null_count,
+    const std::shared_ptr<Buffer>& null_bitmap)
+    : BinaryArray(BINARY, length, offsets, values, null_count, null_bitmap) {}
+
+BinaryArray::BinaryArray(const TypePtr& type, int32_t length,
+    const std::shared_ptr<Buffer>& offsets, const ArrayPtr& values, int32_t null_count,
+    const std::shared_ptr<Buffer>& null_bitmap)
+    : ListArray(type, length, offsets, values, null_count, null_bitmap),
+      bytes_(std::dynamic_pointer_cast<UInt8Array>(values).get()),
```

Comment:
If this cast fails, the result is null. I'm not sure the best way to deal with this; DCHECK maybe?